### ISSUE TITLE
Service wrappers for spring boot application

### ISSUE THREAD

**Issue by wizwjw** [spring-projects/spring-boot#519]
Just like nexus(a maven proxy). Start nexus use `nexus start`,then it print little message, and run as a daemon.

**Comment by DeezCashews** [comment]
I would like this as well.  Currently I am using Apache Commons Daemon to run a web enabled Spring Boot application.  It is configured to run WarLauncher which works perfectly.  My problem is stopping the service gracefully.  I may be missing something, but I don't see any hooks for stopping a running application beyond killing the process.  I am using JVM mode which gives me access to the same instance of WarLauncher that I used to start my application, I just need a "public static void stop" method or ability to pass a "stop" param to the main method.

**Comment by dsyer** [comment]
You can use kill and a pidfile. I think that's what most service wrappers do. Or if you want a Java API you can grab the `ConfigurableApplicationContext` from `SpringApplication.run()` and call `close()` on it. There's no need for the framework to provide anything special for that I think.

**Comment by DeezCashews** [comment]
Yes I'd like to use the Java API since I am using Apache Commons Daemon procrun.  So here is where it gets tricky.  `org.springframework.boot.loader.WarLauncher` creates a special classloader to allow for in place execution similar to a application container.  Its main method calls `new WarLauncher().launch(args)` which sets up the classloader to load WEB-INF/lib, etc. and finally calls your startup class, which is where I simply call `SpringApplication.run(AppConfig.class, args);`  I tried exactly what you described but I can't save the `context` because `WarLauncher` creates a new instance of itself everytime and thus a whole new classloader.  If `WarLauncher` kept the classloader as a static variable I think your method would work.  I could easily write a wrapper class for `WarLauncher` instead of using it directly but unfortunately Spring Boot Maven Plugin's `spring-boot:repackage` goal doesn't have any options for adding classes to the "parent classloader".

**Comment by DeezCashews** [comment]
While this is just for `WarLauncher` projects, the idea could easily be translated to `JarLauncher` and `PropertiesLauncher`.  The idea is to have something that allows one to enter the VM again and be able to get a handle to classes loaded before.  I haven't tested (I tried, but getting the `spring-boot-maven-plugin` to use my new code proved difficult), but I'm 99% sure this would fix my issue and help with issue #644:

```java
public class SharedLoaderWarLauncher extends WarLauncher {

	protected static SharedLoaderWarLauncher INSTANCE;
	
	protected ClassLoader classLoader;
	
	public SharedLoaderWarLauncher() {
		super();
	}
	
	SharedLoaderWarLauncher(Archive archive) {
		super(archive);
	}
	
	@Override
	protected synchronized ClassLoader createClassLoader(URL[] urls) throws Exception {
		if (classLoader == null) {
			classLoader = super.createClassLoader(urls);
		}
		return classLoader;
	}
	
	public static final SharedLoaderWarLauncher getInstance() {
		if (INSTANCE == null) {
			INSTANCE = new SharedLoaderWarLauncher();
		}
		return INSTANCE;
	}

	public static void main(String[] args) {
		getInstance().launch(args);
	}
}
```

**Comment by philwebb** [comment]
https://github.com/rburgst/spring-boot-initscript

**Comment by jschappet** [comment]
I'm using Commons Daemon and have the following stop method working for my application.

	public static void stop(String[] args) {
		try {

			System.out.println("Stopping");

			((ConfigurableApplicationContext) context).close();
			System.out.println("closed context");
		} catch (final Exception e) {
			// handle exceptions properly here
			System.err.println(e.getLocalizedMessage());
			e.printStackTrace();

		}
	}

**Comment by DeezCashews** [comment]
James, That works fine as long as you don't use spring-boot-plugin to
create an executable uber-jar for you.  When it does this, it uses
WarLauncher, JarLauncher, or PropertiesLauncher as its main class and your
application is then invoked from there.  I am assuming you are not using
the spring-boot-plugin if this is working for you, correct?

--
Andrew C Wynham
Principal Consultant
TILT Development, Inc.
703.864.0247

**Comment by jschappet** [comment]
Correct I'm using Apache Commons Daemon procrun and not using spring-boot-plugin.

**Comment by sbarlabanov** [comment]
Why not to use JNA https://github.com/twall/jna?
Just write something like WinServiceLauncher similar to JarLauncher.
We are currently using JNA to start/stop the whole Glassfish domain infrastructure using Windows services (we wrote our own Service Wrapper for Glassfish using JNA). JNA is quite stable, works fine and gives you full control over Windows Service API in Java and has Apache 2 license.

**Comment by DeezCashews** [comment]
I'm sure that'd work fine.  As does using Apache Commons Daemon with the
patch I submitted which took less work than writing a brand new Launcher
(simply adds 5-10 lines to the 3 existing launchers).  I like Commons
Daemon because it is familiar to anyone who has used Tomcat and already
exposes a nice GUI for common items such as JVM args, application args,
manipulating service properties, etc.  Its nice to have a build that
produces a single jar and that jar can either be ran from command line or
started with a service wrapper without modification.  I will look at JNA in
the future though should I need to start/stop additional services from my
application, I can see that being very useful.

**Comment by wimdeblauwe** [comment]
+1 for an easy way to create windows and unix services from a spring boot app

**Comment by jordanjennings** [comment]
Just want to throw in that some of us don't have root access to our deployment boxes (large corporate environment), so although the new init.d stuff in 1.3.0 looks really nice, it's unfortunately not an option for all users (or is there a way to use it without root? not a linux guru here).

I'd love to see a built in way to generate standalone script(s) for startup, shutdown, restart, and status that could be used by Jenkins. If there's already something that does already this please educate me.

**Comment by lokkju** [comment]
the init.d stuff in 1.3.0 solves the problem entirely for me.

**Comment by jordanjennings** [comment]
I just discovered that you can use the built-in init scripts in the jar without needing to link to /etc/init.d. Awesome! The only slightly awkward thing is that you have to set environment variables as described here:

http://docs.spring.io/spring-boot/docs/current-SNAPSHOT/reference/htmlsingle/#_running_a_jar_as_a_regular_not_service_script

So, for example, by setting MODE=service you can do ./myapp.jar start|stop|restart|status

This is very convenient, and what would make it even better is some ability to set the defaults for these properties somewhere (in the build file I'd guess?) so that the jar is pre-built exactly the way that's needed (while still having the ability to override with environment variables if needed).

What I have done for now is created a very simple wrapper that looks something like this:

```
export MODE=service
export PID_FOLDER=.
export LOG_FOLDER=.

./myapp.jar $1
```

This allows ./wrapper.sh start|stop|restart|status to work. Only thing is the naming of the pid and log files isn't ideal, but it gets the job done.

TLDR; providing some way to configure the default values for mode, PID folder/name, and log folder/name in the embedded script would make this feature perfect for my use case.

**Comment by Robin479** [comment]
There's more to a linux daemon than just an init.d interface (start|stop|status.. script).
A daemon process should be started as root (eg. to be able to bind to system port 80), but it should drop to user privileges before initializing the actual application and subsystems – for security reasons. In order to run privileged (root) and unprivileged (user) code within the _same_ JVM instance, a running process has to drop its initial superuser privileges at runtime. That's a native system call, which cannot naively be implemented in Java. It's not impossible to do, but it's certainly a hard nut to crack. The "commons-daemon" package in conjunction with "jsvc" supports it, but you're required to implement their two-stage entry interface (Daemon.initialize()/Daemon.start()) at the expense of a regular main() method  - and you have to invoke "jsvc" instead of "java".

**Comment by dsyer** [comment]
There's no reason as far as I know that it has to be implemented in Java is there? We have a shell script that does it currently, and it runs the app as the user who owns the jar file. Maybe you could have a look and see if it can be improved (https://github.com/spring-projects/spring-boot/blob/master/spring-boot-tools/spring-boot-loader-tools/src/main/resources/org/springframework/boot/loader/tools/launch.script)?

@jordanjennings you can be as flexible as you like if you want to write your own script (e.g. in maven set the `embeddedLaunchScript` property for the repackage goal). To change the default mode in the default script you can use `embeddedLaunchScriptProperties`. If you want to change other defaults (they seem sensible to me), you could send a PR.

**Comment by jordanjennings** [comment]
If any changes are made to start this as root please also allow a non-root option. I'm guessing that in many cases applications will be running behind a web server and will be running on 8080 or some other high numbered port, so root is not needed to start. Assuming that's the case, root might be better as an opt-in rather than an opt-out (for only when you need to bind to port < 1024).

@dsyer Thanks for pointing out `embeddedLaunchScriptProperties`. The defaults are very sensible for when you're actually using it via init.d, but on our boxes we don't have root and don't have access to /var/log or /var/run, so we're using this as a standalone script which is why some minor tweaks are needed. Being able to set them using `embeddedLaunchScriptProperties` would be perfect, so I'll look into sending a PR to add more to that.

**Comment by Polve** [comment]
Speaking about integration scripts, a configuration file for systemd (used on most modern linux distros) would be very useful too :)

**Comment by dsyer** [comment]
@jordanjennings the script that we provide runs as the current user, unless that user is root, in which case it tries to switch to the owner of the jar file. Seems pretty sensible to me, and leaves control in the hands of the caller, but if you have something more sophisticated in mind we can consider it (PR more than welcome).

@Polve as far as I know systemd is not used in Ubuntu (yet). I assume something like a systemd config file would be part of an rpm (or other UN*X package), and since we have no explicit support for building packages other than JARs, it would seem like there isn't really any way to add a new feature to Spring Boot. Anyway, if you have a proposal for a config file and something that needs to be done with it during a build process, please explain in a bit more detail (or preferably send a PR).

**Comment by Robin479** [comment]
@dsyer Starting and running an app as an unprivileged user is different from starting an app as root and then dropping the running process to unprivileged user permissions. The latter allows the process to access/acquire protected system resources such as low port binding and loading of read-protected files, but it also makes sure that such privileged actions cannot be executed by the actual user app where they might harm the system. This would, of course, require some major work on the Java side of the init.d feature, too, because the needed system resources would have to be identified within the app, and acquired and configured on the user's behalf – all before the main application context gets initialized (i.e. a two stage application context).
I'm not implying that this should be the way to go. I'm only pointing things out here. The majority of users doesn't need access to protected resources anyway.

**Comment by jordanjennings** [comment]
@dsyer Yes, the current behavior with privileges is exactly what I need. My comment about root was just in reaction to @Robin479 's comment which suggested it should run as root then drop privileges. I don't want root, and since my jars will be run as a non-root user it's all good as-is.

The only piece I might send a pull request for (once I can get some time to test it out) is to add more build-time configuration for the script. Thanks!

**Comment by Polve** [comment]
@dsyer : Since latest release, systemd is the default in both Ubuntu and Debian

**Comment by dsyer** [comment]
This discussion has been great. It's now time to narrow down to specific feature requests. We have the launch.script support already in 1.3. See #3829 for example.

**Comment by tumakha** [comment]
Simple startup scripts with Java version checking for Linux [auth-agent](https://github.com/tumakha/auth-agent/blob/master/src/assembly/bin-linux/auth-agent)
and Windows [auth-agent.bat](https://github.com/tumakha/auth-agent/blob/master/src/assembly/root/bin/auth-agent.bat)
